subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java

```diff
@@ -53,6 +54,8 @@ public SettingsInternal findAndLoadSettings(GradleInternal gradle) {
             return settings;
         }
 
+        deprecateWarningIfNecessary(startParameter, settings);
```

Comment:
Why do we need this additional logic? If we reach this point in the code, shouldn't we always show a warning, since we are retrying and we don't want to retry in the future.

Reply:
Good catch. It's a bit complex:

`spec.containsProject(settings.getProjectRegistry())` not only includes the scenario we're going to deprecate. For example, [BuildFileProjectSpec](https://github.com/gradle/gradle/blob/c8143c9bcf412a2d19053b8c926c7787c1dc8087/subprojects/core/src/main/java/org/gradle/initialization/BuildFileProjectSpec.java#L43) checks if `buildFile` in start parameter matches the project's default build file. IOW, if we don't add the additional logic, some scenarios where deprecation warning shouldn't appear like [this test](https://github.com/gradle/gradle/blob/eadb9804831b7d9f8ed38b34ac6522931ea70f75/subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/ProjectLoadingIntegrationTest.java#L258) would be broken:

```
@Test
public void noDeprecationWarningAppearsWhenOutsideBuildUsesAnotherBuildFile() {
    testFile("settings.gradle").write("include 'another'");
    TestFile renamedBuildGradle = getTestDirectory().file("renamed_build.gradle").createFile();

    usingBuildFile(renamedBuildGradle).inDirectory(getTestDirectory()).withSearchUpwards().withTasks("help").run();
}
```

In this test, we simply change the build file's name, this will make `spec.containsProject(settings.getProjectRegistry())` return false. The additional logic aims at avoiding deprecation warnings that shouldn't appear.

Reply:
Seems like the original logic is wrong here. When you pass the `-b` option, `settings.gradle` shouldn't be evaluated at all.

Reply:
Why so? `settings.gradle` defines the project structure, it's independent from what main build file you are using (`-b`)

Reply:
@oehme What's more, there's another weird test case which forced me to add some extra logic: 

```
TestFile settingsDir = testFile("gradle");
TestFile settingsFile = settingsDir.file("settings.gradle");
settingsFile.writelns(
                "rootProject.projectDir = new File(settingsDir, '../root')"
        );
getTestDirectory().createDir("root").file("build.gradle").writelns("task thing");

inTestDirectory().withArguments("-p", settingsDir.getAbsolutePath()).withTasks("thing").runWithFailure()
                .assertHasDescription("Task 'thing' not found in root project 'gradle'.");
```

Should we correct it as well?

https://github.com/gradle/gradle/blob/efbdec90600469a9de04978f6ec40b128e4a3d21/subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/ProjectLoadingIntegrationTest.java#L314

Reply:
@melix the `-b` option ignores settings.gradle, see the documentation.

Reply:
@blindpirate `settingsFileGetsIgnoredWhenUsingSettingsOnlyDirectoryAsProjectDirectory` seems wrong, can you find out why we added this? Was there a bug report or feature request or some trace in the git history?

Reply:
@oehme I tried but failed. It's in https://github.com/gradle/gradle/commit/e514c095ebaef339f328012cfa1feffea7ea3d3e with only commit message `+review REVIEW-5066` but I can't login `JIRA`. It looks like a feature story rather than a bug fix. Anyway, I'll try to polish it up.

Reply:
@melix @oehme I've created https://github.com/gradle/gradle/issues/3227 to track `-b` issue. For now, I'd say let's keep this behavior until we decide how to deal with the `-b`/`-p`/`-c` issue.